test/helpers/spy_input.cc

```diff
@@ -57,12 +56,19 @@ const char * SpyInput::read(void *payload, uint32_t *bytes_read) {
    * This class stores its entire `content` in a contiguous buffer, but we want
    * to ensure that the code under test cannot accidentally read more than
    * `*bytes_read` bytes past the returned pointer. To make sure that this type
-   * of error does not fly, we copy the chunk into a zeroed-out buffer and
+   * of error does not fly, we allocate a separate buffer for each request and
    * return a reference to that buffer, rather than a pointer into the main
-   * content.
+   * content. The temporary buffer only fits `*bytes_read` bytes so valgrind
+   * can detect code reading too many bytes from the buffer.
    */
-  memset(spy->buffer, 0, spy->buffer_size);
-  memcpy(spy->buffer, result.data(), byte_count);
+  delete[] spy->buffer;
+  if (byte_count) {
+    spy->buffer = new char[byte_count];
+    memcpy(spy->buffer, result.data(), byte_count);
+  } else {
+    spy->buffer = nullptr;
+  }
+
```

Comment:
The cost of `new` and `delete` get lost in the noise afaict.

With `master`:
```
$ git rev-parse HEAD
afb499bf2e5c9aee78ae9a8611268008e1136359
$ for x in $(echo "1 2 3 4 5"); do time ./out/Test/tests >/dev/null; done
real    0m20.394s
user    0m20.068s
sys     0m0.148s

real    0m21.864s
user    0m21.492s
sys     0m0.168s

real    0m20.518s
user    0m20.136s
sys     0m0.168s

real    0m20.215s
user    0m19.908s
sys     0m0.148s

real    0m20.478s
user    0m20.116s
sys     0m0.192s
```

this branch:
```
$ git rev-parse HEAD
52cec9ed39a56f226d47499db6362a192138fd5a
$ for x in $(echo "1 2 3 4 5"); do time ./out/Test/tests >/dev/null; done

real    0m19.423s
user    0m19.084s
sys     0m0.160s

real    0m19.946s
user    0m19.588s
sys     0m0.188s

real    0m21.006s
user    0m20.672s
sys     0m0.172s

real    0m19.634s
user    0m19.304s
sys     0m0.160s

real    0m19.492s
user    0m19.188s
sys     0m0.128s
```